overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,373 @@
+		st.Unlock()
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	switch resp.StatusCode {
+	case 200, 201:
+	case 202:
+		return nil, errPoll
+	default:
+		st.Lock()
+		t.Errorf("cannot POST request for a serial: unexpected status %d", resp.StatusCode)
+		st.Unlock()
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	// decode body with serial assertion
+	dec := asserts.NewDecoder(resp.Body)
+	got, err := dec.Decode()
+	if err != nil { // assume broken i/o
+		st.Lock()
+		t.Errorf("cannot read response to request for a serial: %v", err)
+		st.Unlock()
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	serial, ok := got.(*asserts.Serial)
+	if !ok {
+		return nil, fmt.Errorf("request for a serial response did not return a serial assertion")
+	}
+
+	return serial, nil
+}
+
+func (m *DeviceManager) doRequestSerial(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	device, err := auth.Device(st)
+	if err != nil {
+		return err
+	}
+
+	privKey, err := m.keyPair()
+	if err != nil {
+		return err
+	}
+
+	// make this idempotent, look if we have already a serial assertion
+	// for privKey
+	serials, err := assertstate.DB(st).FindMany(asserts.SerialType, map[string]string{
+		"brand-id":      device.Brand,
+		"model":         device.Model,
+		"device-key-id": privKey.PublicKey().ID(),
+	})
+
+	if len(serials) == 1 {
+		// means we saved the assertion but didn't get to the end of the task
+		device.Serial = serials[0].(*asserts.Serial).Serial()
+		auth.SetDevice(st, device)
+		t.SetStatus(state.DoneStatus)
+		return nil
+	}
+	if len(serials) > 1 {
+		return fmt.Errorf("internal error: multiple serial assertions for the same device key")
+	}
+
+	client := &http.Client{Timeout: 30 * time.Second}
+
+	var serialSup serialSetup
+	err = t.Get("serial-setup", &serialSup)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	// NB: until we get at least an Accepted (202) we need to
+	// retry from scratch creating a new request-id because the
+	// previous one used could have expired
+
+	if serialSup.SerialRequest == "" {
+		serialRequest, err := prepareSerialRequest(t, privKey, device, client)
+		if err != nil { // errors & retries
+			return err
+		}
+
+		serialSup.SerialRequest = serialRequest
+	}
+
+	serial, err := submitSerialRequest(t, serialSup.SerialRequest, client)
+	if err == errPoll {
+		// we can/should reuse the serial-request
+		t.Set("serial-setup", serialSup)
+		t.Logf("polling for a serial in 60 seconds")
+		return &state.Retry{After: 60 * time.Second}
+	}
+	if err != nil { // errors & retries
+		return err
+	}
+
+	// TODO: (possibly refetch brand key and)
+	// TODO: double check brand, model and key hash
+
+	// add the serial assertion to the system assertion db
+	err = assertstate.Add(st, serial)
+	if err != nil {
+		return err
+	}
+
+	device.Serial = serial.Serial()
+	auth.SetDevice(st, device)
+	t.SetStatus(state.DoneStatus)
```

Comment:
returning does the same but through a unlock/relock window, so we could crash in the middle there, for this one is not a problem because anyway it checks for the saved assertion to be idempotent;

but it's a good habit, in snapstate there are places that would need more logic of the kind "did I already update SnapState or not?" if they didn't do this